cli/flox-watchdog/src/logger.rs

```diff
@@ -40,3 +48,245 @@ pub(crate) fn init_logger(file_path: &Option<PathBuf>) -> Result<(), anyhow::Err
+/// `log_heartbeat`.
+fn gc_logs_watchdog(dir: impl AsRef<Path>, keep_days: u64) -> Result<()> {
+    let mut files = glob_log_files(dir, "watchdog.*.log")?;
+    let threshold = duration_from_days(keep_days);
+    let now = SystemTime::now();
+
+    // Defaults to keeping if mtime is not supported by platform or filesystem.
+    files.retain(|file| file_older_than(file, now, threshold).unwrap_or(false));
+
+    for file in files {
+        try_delete_log(file);
+    }
+
+    Ok(())
+}
+
+/// Garbage collects services log files, keeping the last N files by filename.
+/// This relies on the log files having a timestamp in their filename.
+fn gc_logs_services(dir: impl AsRef<Path>, keep_last: usize) -> Result<()> {
+    let mut files = glob_log_files(dir, "services.*.log")?;
+    if files.len() <= keep_last {
+        return Ok(());
+    }
+
+    files.sort_unstable();
+    if files.len() > keep_last {
+        files.truncate(files.len() - keep_last);
+    }
+
+    for file in files {
+        try_delete_log(file);
+    }
+
+    Ok(())
+}
+
+/// Construct a Duration from `days`.
+const fn duration_from_days(days: u64) -> Duration {
+    let seconds_in_day = 24 * 60 * 60;
+    Duration::from_secs(days * seconds_in_day)
+}
+
+/// Returns whether a file is older than `threshold`.
+fn file_older_than(file: impl AsRef<Path>, now: SystemTime, threshold: Duration) -> Result<bool> {
+    let metadata = file
+        .as_ref()
+        .metadata()
+        .context("failed to get file metadata")?;
+    let modified = metadata.modified()?;
+    let elapsed = now.duration_since(modified)?;
+    let older = elapsed > threshold;
+
+    Ok(older)
+}
+
+/// Glob files matching a pattern, ignoring any unreadable paths.
+fn glob_log_files(dir: impl AsRef<Path>, name: &str) -> Result<Vec<PathBuf>> {
+    let pattern = format!("{}/{name}", dir.as_ref().to_string_lossy());
+    let paths = glob(&pattern).context("failed to glob logs")?;
+    let mut files: Vec<PathBuf> = paths.filter_map(Result::ok).collect();
+    files.retain(|path| path.is_file());
+    Ok(files)
+}
+
+/// Delete a log file. Errors are traced and not bubbled up, so that we don't
+/// get stuck on individual files.
+fn try_delete_log(file: impl AsRef<Path>) {
+    remove_file(file.as_ref()).unwrap_or_else(|err| error!(%err, "failed to delete log"));
+    debug!(path = %file.as_ref().display(), "deleted log");
+}
+
+#[cfg(test)]
+mod tests {
+    use std::fs::{create_dir, File};
+
+    use filetime::{set_file_mtime, FileTime};
+    use tempfile::tempdir;
+
+    use super::*;
+
+    /// Create a test log file. Optionally set a modified time in days.
+    fn create_log_file(dir: &Path, name: &str, days: Option<u64>) -> PathBuf {
+        let path = dir.join(name);
+        File::create(&path).unwrap();
+
+        if let Some(days) = days {
+            let mtime = SystemTime::now() - duration_from_days(days);
+            set_file_mtime(&path, FileTime::from_system_time(mtime)).unwrap();
```

Comment:
suggestion: replace `filetime` for `std` method
```suggestion
            let mtime = SystemTime::now() - duration_from_days(days);
            File::open(path).unwrap().set_modified(mtime).unwrap();
```

Reply:
Done (with `&path` because we return it) 👍🏻